cpp/src/arrow/dataset/scanner.cc

```diff
@@ -60,19 +61,18 @@ static ScanTaskIterator GetScanTaskIterator(DataFragmentIterator fragments,
 }
 
 static ScanTaskIterator ProjectAndFilterScanTaskIterator(
-    ScanTaskIterator it, std::shared_ptr<Expression> filter,
+    ScanTaskIterator it, ExpressionPtr filter,
     std::shared_ptr<ExpressionEvaluator> evaluator,
     std::shared_ptr<RecordBatchProjector> projector) {
   // Wrap the scanner ScanTask with a FilterAndProjectScanTask
-  auto wrap_scan_task = [filter, evaluator, projector](
-                            std::unique_ptr<ScanTask> task) -> std::unique_ptr<ScanTask> {
+  auto wrap_scan_task = [filter, evaluator, projector](ScanTaskPtr task) -> ScanTaskPtr {
     return internal::make_unique<FilterAndProjectScanTask>(std::move(task), filter,
```

Comment:
it looks a bit funny to have `make_unique` here since you're return type is `shared_ptr`, but it's not at all incorrect

Reply:
Will, fix.